src/cockpit/channels/filesystem.py

```diff
@@ -41,53 +48,189 @@ def tag_from_path(path):
+    fd: Optional[Handle] = None
+    exists: Optional[bool] = None
+    path: str
+
+    def event(self, event: str, name: str) -> None:
+        attrs: JsonObject = {'event': event, 'path': name}
+
+        if event in ('created', 'attribute-changed', 'present', 'done-hint'):
+            assert self.fd is not None
+
+            try:
+                attrs.update(self.reporter.report(name, dir_fd=self.fd))
+            except FileNotFoundError:
+                return  # we'll get the delete event delivered to us soon...
+            except OSError:
+                pass  # we're just not going to report attributes in that case
+
+        logger.debug('%s %s', self.channel, attrs)
+        self.send_text(json.dumps(attrs) + '\n')
+
+    def do_close(self, error: Optional[int] = None) -> None:
+        if self.watch is not None:
+            self.watch.close()
+
+        if error is None:
+            self.close()
+        elif error == errno.ENOENT:
+            self.close(problem='not-found', message=os.strerror(error))
+        elif error in (errno.EPERM, errno.EACCES):
+            self.close(problem='access-denied', message=os.strerror(error))
         else:
-            mode = 'special'
+            self.close(problem='internal-error', message=os.strerror(error))
+
+    def do_inotify_event(self, mask: InotifyEvent, _cookie: int, name: Optional[bytes]) -> None:
+        # If self.children is True, return if name is None
+        # If self.children is False, return if name is non-None
+        if self.children == (name is None):
+            return
+
+        if mask & (InotifyEvent.MOVED_FROM | InotifyEvent.DELETE):
+            event = 'deleted'
+        elif mask & (InotifyEvent.MOVED_TO | InotifyEvent.CREATE):
+            event = 'created'
+        elif mask & InotifyEvent.ATTRIB:
+            event = 'attribute-changed'
+        elif mask & InotifyEvent.CLOSE_WRITE:
+            event = 'done-hint'
+        else:
+            event = 'changed'
+
+        self.event(event, name.decode('utf-8') if name is not None else self.path)
 
-        self.send_message(event=event, path=entry.name, type=mode)
+    def do_identity_changed(self, fd: Optional[Handle], err: Optional[int]) -> None:
+        self.fd = fd
+
+        if self.children and err is not None:
+            self.do_close(err)
+        elif (self.fd is not None) != self.exists:
+            if self.exists is not None:
+                self.event('deleted' if fd is None else 'created', self.path)
+            self.exists = self.fd is not None
 
     def do_open(self, options: JsonObject) -> None:
-        path = get_str(options, 'path')
-        watch = get_bool(options, 'watch', default=True)
+        self.path = get_str(options, 'path')
 
-        if watch:
-            raise ChannelError('not-supported', message='watching is not implemented, use fswatch1')
+        try:
+            self.reporter = AttributeReporter(get_strv(options, 'attrs', self.attrs))
```

Comment:
Wrt. the above comment, attrs does not get filtered for validity here.

Reply:
ya, indeed.  we catch AttributeError here and send `not-supported` problem.

The output looks like:

```
⬢[lis@toolbox fslist1]$ cpf open fswatch1 path=/ "attrs=['q']" | spy -m cockpit.bridge
...
{
  "channel": "ch1",
  "command": "close",
  "message": "'AttributeReporter' object has no attribute 'do_get_q'",
  "problem": "not-supported"
}
```

Reply:
Ack thanks, that wasn't obvious from the patch.